File: libs/adsp-service-net-sdk/Access/TokenProvider.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using RestSharp;

namespace Adsp.Sdk.Access;

internal sealed class TokenProvider : ITokenProvider, IDisposable
{
  private readonly Lock _lock = new();
  private string? _token;
  private DateTime _expiry;

  private readonly ILogger<TokenProvider> _logger;
  private readonly Uri _authUrl;
  private readonly IRestClient _client;
  private readonly string _clientId;
  private readonly string _clientSecret;

  public TokenProvider(ILogger<TokenProvider> logger, IOptions<AdspOptions> options, IRestClient? client = null)
  {
    if (options.Value.AccessServiceUrl == null)
    {
      throw new ArgumentException("Provided options must include value for AccessServiceUrl.", nameof(options));
    }

    if (options.Value.ServiceId == null)
    {
      throw new ArgumentException("Provided options must include value for ServiceId.", nameof(options));
    }

    if (options.Value.ClientSecret == null)
    {
      throw new ArgumentException("Provided options must include value for ClientSecret.", nameof(options));
    }

    _logger = logger;
    _authUrl = new Uri(
      options.Value.AccessServiceUrl,
      $"/auth/realms/{options.Value.Realm ?? AccessConstants.CoreRealm}/protocol/openid-connect/token"
    );
    _client = client ?? new RestClient(_authUrl);
    _clientId = options.Value.ServiceId.ToString();
    _clientSecret = options.Value.ClientSecret;
  }

  private string? GetCachedToken(out DateTime expiry)
  {
    lock (_lock)
    {
      expiry = _expiry;
      return _token;
    }
  }

  private void SetCachedToken(string token, DateTime expiry)
  {
    lock (_lock)
    {
      _expiry = expiry;
      _token = token;
    }
  }

  public async Task<string> GetAccessToken()
  {
    var token = GetCachedToken(out DateTime expiry);
    if (!String.IsNullOrEmpty(token) && DateTime.Now < expiry)
    {
      _logger.LogDebug("Using existing access token...");
      return token;
    }
    else
    {
      _logger.LogDebug("Requesting access token from {AuthUrl}...", _authUrl);

      try
      {
        var request = new RestRequest("", Method.Post)
          .AddParameter("grant_type", "client_credentials")
          .AddParameter("client_id", _clientId, ParameterType.GetOrPost)
          .AddParameter("client_secret", _clientSecret, ParameterType.GetOrPost);

        var response = await _client.PostAsync<TokenResponse>(request);
        if (!String.IsNullOrEmpty(response?.AccessToken))
        {
          token = response.AccessToken;
          SetCachedToken(token, DateTime.Now.AddSeconds(response.ExpiresIn - 60));

          _logger.LogDebug("Retrieved and cached access token.");
        }
      }
      catch (Exception e)
      {
        _logger.LogError(e, "Error encountered retrieving access token");
        throw;
      }

      return token!;
    }
  }

  public void Dispose()
  {
    _client.Dispose();
  }
}

File: libs/adsp-service-flask-sdk/adsp_service_flask_sdk/token_provider.py
from datetime import datetime, timedelta
import logging
from operator import itemgetter
from threading import Lock
from typing import Optional

from httpx import RequestError, post

from .adsp_id import AdspId


class TokenProvider:
    _logger = logging.getLogger("adsp.token-provider")
    __response_values_getter = itemgetter("access_token", "expires_in")

    def __init__(
        self,
        access_service_url: str,
        service_id: AdspId,
        client_secret: str,
        realm="core",
    ) -> None:
        self.__token: Optional[str] = None
        self.__expires_at: Optional[datetime] = None
        self.__token_lock = Lock()
        self.__access_service_url = access_service_url
        self.__realm = realm
        self.__service_id = service_id
        self.__client_secret = client_secret

    def _get_cached_token(self) -> Optional[str]:
        with self.__token_lock:
            if (
                self.__token
                and self.__expires_at
                and datetime.now() < self.__expires_at
            ):
                return self.__token
            else:
                return None

    def _set_cached_token(self, token: str, expires_at: datetime) -> None:
        with self.__token_lock:
            self.__token = token
            self.__expires_at = expires_at

    def get_access_token(self) -> Optional[str]:
        token = self._get_cached_token()

        if token is None:
            try:
                request_data = {
                    "grant_type": "client_credentials",
                    "client_id": str(self.__service_id),
                    "client_secret": self.__client_secret,
                }
                result = post(
                    f"{self.__access_service_url}/auth/realms/{self.__realm}/protocol/openid-connect/token",
                    data=request_data,
                ).json()

                access_token, expires_in = self.__response_values_getter(result)
                if type(access_token) is str and type(expires_in) is int:
                    self._set_cached_token(
                        access_token,
                        datetime.now() + timedelta(seconds=(expires_in - 60)),
                    )

                return access_token
            except RequestError as err:
                self._logger.error("Error encountered retrieving access token. %s", err)
                raise

        return token
